Sources/AsyncAlgorithms/Channels/ChannelStateMachine.swift

```diff
@@ -0,0 +1,367 @@
+    }
+  }
+
+  enum NextIsSuspendedAction {
+    case suspend
+    case resumeConsumer(element: Element?)
+    case resumeConsumerWithError(error: Error)
+    case resumeProducerAndConsumer(continuation: UnsafeContinuation<Void, Never>?, element: Element?)
+  }
+
+  mutating func nextIsSuspended(
+    continuation: UnsafeContinuation<Element?, any Error>,
+    consumerId: Int
+  ) -> NextIsSuspendedAction {
+    switch self.state {
+      case .channeling(
+        let suspendedProducers,
+        let cancelledProducers,
+        var suspendedConsumers,
+        var cancelledConsumers
+      ) where suspendedProducers.isEmpty:
+        // we are idle or waiting for producers
+        let suspendedConsumer = SuspendedConsumer(id: consumerId, continuation: continuation)
+        if let _ = cancelledConsumers.remove(suspendedConsumer) {
+          // the consumer was already cancelled, we resume it
+          self.state = .channeling(
+            suspendedProducers: suspendedProducers,
+            cancelledProducers: cancelledProducers,
+            suspendedConsumers: suspendedConsumers,
+            cancelledConsumers: cancelledConsumers
+          )
+          return .resumeConsumer(element: nil)
+        }
+
+        // we stack the incoming consumer in a suspended state
+        suspendedConsumers.append(suspendedConsumer)
+        self.state = .channeling(
+          suspendedProducers: suspendedProducers,
+          cancelledProducers: cancelledProducers,
+          suspendedConsumers: suspendedConsumers,
+          cancelledConsumers: cancelledConsumers
+        )
+        return .suspend
+
+      case .channeling(var suspendedProducers, let cancelledProducers, let suspendedConsumers, var cancelledConsumers):
+        // we are waiting for consumers
+        let suspendedConsumer = SuspendedConsumer.placeHolder(id: consumerId)
+        if let _ = cancelledConsumers.remove(suspendedConsumer) {
+          // the consumer was already cancelled, we resume it
+          self.state = .channeling(
+            suspendedProducers: suspendedProducers,
+            cancelledProducers: cancelledProducers,
+            suspendedConsumers: suspendedConsumers,
+            cancelledConsumers: cancelledConsumers
+          )
+          return .resumeConsumer(element: nil)
+        }
+        // we resume the first producer
+        let suspendedProducer = suspendedProducers.removeFirst()
+        self.state = .channeling(
+          suspendedProducers: suspendedProducers,
+          cancelledProducers: cancelledProducers,
+          suspendedConsumers: suspendedConsumers,
+          cancelledConsumers: cancelledConsumers
+        )
+        return .resumeProducerAndConsumer(
+          continuation: suspendedProducer.continuation,
+          element: suspendedProducer.element
+        )
+
+      case .terminated(.finished):
+        return .resumeConsumer(element: nil)
+
+      case .terminated(.failed(let error)):
```

Comment:
I don't think that this is aligned with the rule that `next()` should throw an error once and after that it should return `nil`. This code right now throws the same error over and over again. Implementing the expected behaviour is going to get interesting since it means that we have to keep track of each iterator and what they saw. Maybe we can achieve this by creating a single ID for an iterator that we keep inside the iterator and use for every next() call on the state machine instead of generating an ID per next call.

Reply:
It is not exactly what happens. When the state is .finished(.failure) and there are some suspended consumers then they all receive the failure and the state is set to  .terminated(.finished). The subsequent call to next will receive “nil”.

Reply:
@FranzBusch, @phausler before making any changes I'd like to be sure of the expected behaviour. I think that what I have now is aligned with the previous implementation of `AsyncChannel`.

Reply:
Maybe I am misreading the code here as well. If the behaviour is the same then it should be good!

Reply:
this is the unit test that checks that:

```swift
  func test_asyncThrowingChannel_resumes_producers_and_discards_additional_elements_when_fail_is_called() async throws {
 // Given: an AsyncThrowingChannel
 let sut = AsyncThrowingChannel<Int, Error>()

 // Given: 2 suspended send operations
 let task1 = Task {
   await sut.send(1)
 }

 let task2 = Task {
   await sut.send(2)
 }

 // When: failing the channel
 sut.fail(Failure())

 // Then: the send operations are resumed
 _ = await (task1.value, task2.value)

 // When: sending an extra value
 await sut.send(3)

 // Then: the send operation is resumed
 // Then: the iteration is resumed with a failure
 var collected = [Int]()
 do {
   for try await element in sut {
     collected.append(element)
   }
 } catch {
   XCTAssertTrue(collected.isEmpty)
   XCTAssertEqual(error as? Failure, Failure())
 }

 // When: requesting a next value
 var iterator = sut.makeAsyncIterator()
 let pastFailure = try await iterator.next()

 // Then: the past failure is nil
 XCTAssertNil(pastFailure)
}
```

Reply:
Thanks for posting this. Do we also have a test that sets up an iterator `next()` call before and then calls `sut.fail()`?

Reply:
yep

```swift
func test_asyncThrowingChannel_resumes_consumers_when_fail_is_called() async throws {
  // Given: an AsyncThrowingChannel
  let sut = AsyncThrowingChannel<Int, Error>()

  // Given: 2 suspended iterations
  let task1 = Task<Int?, Error> {
    var iterator = sut.makeAsyncIterator()
    return try await iterator.next()
  }

  let task2 = Task<Int?, Error> {
    var iterator = sut.makeAsyncIterator()
    return try await iterator.next()
  }

  // When: failing the channel
  sut.fail(Failure())

  // Then: the iterations are resumed with the error
  do {
    _ = try await (task1.value, task2.value)
  } catch {
    XCTAssertEqual(error as? Failure, Failure())
  }

  // When: requesting a next value
  var iterator = sut.makeAsyncIterator()
  let pastFailure = try await iterator.next()

  // Then: the past failure is nil
  XCTAssertNil(pastFailure)
}
```

Reply:
Hm that's not really testing the same. I would like to see that test but with the same iterator calling next twice. Like this

```
  let task1 = Task<Int?, Error> {
    var iterator = sut.makeAsyncIterator()
    try await iterator.next() // Need to catch the values of both next and return them in an array to assert
    try await iterator.next()
  }
```

Reply:
ok. 

Is that ok?

```
  func test_asyncThrowingChannel_resumes_consumer_when_fail_is_called() async throws {
    // Given: an AsyncThrowingChannel
    let sut = AsyncThrowingChannel<Int, Error>()

    // Given: suspended iteration
    let task = Task<Int?, Error> {
      var iterator = sut.makeAsyncIterator()

      do {
        _ = try await iterator.next()
      } catch {
        XCTAssertEqual(error as? Failure, Failure())
      }

      return try await iterator.next()
    }

    // When: failing the channel
    sut.fail(Failure())

    // Then: the iterations are resumed with the error and the next element is nil
    do {
      let collected = try await task.value
      XCTAssertNil(collected)
    } catch {
      XCTFail("The task should not fail, the past failure element should be nil, not a failure.")
    }
  }
```

Reply:
I think almost, just one slight change
```
  func test_asyncThrowingChannel_resumes_consumer_when_fail_is_called() async throws {
    // Given: an AsyncThrowingChannel
    let sut = AsyncThrowingChannel<Int, Error>()

    // Given: suspended iteration
    let task = Task<Int?, Error> {
      var iterator = sut.makeAsyncIterator()

      do {
        _ = try await iterator.next()
        XCTFail("We expect the above call to throw")
      } catch {
        XCTAssertEqual(error as? Failure, Failure())
      }

      return try await iterator.next()
    }

    // When: failing the channel
    sut.fail(Failure())

    // Then: the iterations are resumed with the error and the next element is nil
    do {
      let collected = try await task.value
      XCTAssertNil(collected)
    } catch {
      XCTFail("The task should not fail, the past failure element should be nil, not a failure.")
    }
  }
````

Would be great if you could add that test

Reply:
I had to do a few tweaks to make it pass ... nice catch.
I've pushed the updated version.